File: sdks/go/environments.go
package fuse

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type transport struct {
	baseURL *url.URL
	http    *http.Client
	// streamHTTP is a no-timeout client used for long-lived SSE
	// streams (events). do not use it for normal requests.
	streamHTTP *http.Client
	bearer     string
	userAgent  string
	requestID  func() string
}

func (t *transport) newRequest(ctx context.Context, method, path string, query url.Values, body any) (*http.Request, error) {
	if t == nil {
		return nil, errors.New("transport is nil")
	}
	if ctx == nil {
		return nil, errors.New("context is nil")
	}
	if !strings.HasPrefix(path, "/") {
		path = "/" + path
	}
	ref, err := url.Parse(path)
	if err != nil {
		return nil, fmt.Errorf("parse path: %w", err)
	}
	if len(query) > 0 {
		ref.RawQuery = query.Encode()
	}
	var buf *bytes.Buffer
	if body != nil {
		buf = &bytes.Buffer{}
		enc := json.NewEncoder(buf)
		enc.SetEscapeHTML(false)
		if err := enc.Encode(body); err != nil {
			return nil, fmt.Errorf("encode body: %w", err)
		}
	}
	base := &url.URL{}
	if t.baseURL != nil {
		base = t.baseURL
	}
	u := base.ResolveReference(ref)
	// use an io.Reader interface var, not a *bytes.Reader, so a
	// missing body is an untyped nil. passing a typed nil *bytes.Reader
	// makes net/http panic when it type-asserts and calls Len().
	var reader io.Reader
	if buf != nil {
		reader = bytes.NewReader(buf.Bytes())
	}
	req, err := http.NewRequestWithContext(ctx, method, u.String(), reader)
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if t.bearer != "" {
		req.Header.Set("Authorization", "Bearer "+t.bearer)
	}
	if t.userAgent != "" {
		req.Header.Set("User-Agent", t.userAgent)
	}
	if t.requestID != nil {
		if id := t.requestID(); id != "" {
			req.Header.Set(requestIDHeader, id)
		}
	}
	return req, nil
}

func (t *transport) do(req *http.Request) (*http.Response, error) {
	if t == nil {
		return nil, errors.New("transport is nil")
	}
	client := t.http
	if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("do request: %w", err)
	}
	return resp, nil
}

// doStream runs a request on the no-timeout client, for calls whose duration is
// bounded by the request context rather than a fixed client timeout (a
// long-running exec, an SSE stream).
func (t *transport) doStream(req *http.Request) (*http.Response, error) {
	if t == nil {
		return nil, errors.New("transport is nil")
	}
	client := t.streamHTTP
	if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("do request: %w", err)
	}
	return resp, nil
}

type EnvironmentsService struct {
	t *transport
}

func newEnvironmentsService(t *transport) *EnvironmentsService {
	return &EnvironmentsService{t: t}
}

type ListEnvironmentsOptions struct {
	TaskID string
	State  string
	HostID string

	// Limit caps the page size (server default 50, max 200; <=0 uses the
	// server default). Only consulted by ListPage — List always requests
	// the server's max page size internally since it walks every page.
	Limit int

	// Cursor resumes after a previous page's EnvironmentPage.NextCursor. It
	// is an opaque token; treat it as such rather than parsing it.
	Cursor string
}

// EnvironmentPage is one page of a List call, plus the cursor to fetch the
// next one.
type EnvironmentPage struct {
	Environments []EnvironmentInfo `json:"environments"`
	// NextCursor is empty once there are no more results.
	NextCursor string `json:"next_cursor,omitempty"`
}

// List returns every environment matching opt, transparently walking every
// result page. For explicit single-page control (e.g. a CLI --cursor
// flag), use ListPage.
func (s *EnvironmentsService) List(ctx context.Context, opt ListEnvironmentsOptions) ([]EnvironmentInfo, error) {
	var out []EnvironmentInfo
	cursor := opt.Cursor
	for {
		page, err := s.ListPage(ctx, ListEnvironmentsOptions{
			TaskID: opt.TaskID,
			State:  opt.State,
			HostID: opt.HostID,
			Limit:  maxPageLimit,
			Cursor: cursor,
		})
		if err != nil {
			return nil, err
		}
		out = append(out, page.Environments...)
		if page.NextCursor == "" {
			break
		}
		cursor = page.NextCursor
	}
	return out, nil
}

// ListPage returns one page of environments matching opt.
func (s *EnvironmentsService) ListPage(ctx context.Context, opt ListEnvironmentsOptions) (*EnvironmentPage, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	values := url.Values{}
	if opt.TaskID != "" {
		values.Set("task_id", opt.TaskID)
	}
	if opt.State != "" {
		values.Set("state", opt.State)
	}
	if opt.HostID != "" {
		values.Set("host_id", opt.HostID)
	}
	setPaginationParams(values, opt.Limit, opt.Cursor)
	req, err := s.t.newRequest(ctx, http.MethodGet, "/v1/environments", values, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var out environmentList
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("decode environments: %w", err)
	}
	page := &EnvironmentPage{Environments: out.Environments}
	if out.NextCursor != nil {
		page.NextCursor = *out.NextCursor
	}
	return page, nil
}

func (s *EnvironmentsService) Get(ctx context.Context, vmID string) (*EnvironmentInfo, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	if vmID == "" {
		return nil, errors.New("vm id is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID)
	req, err := s.t.newRequest(ctx, http.MethodGet, path, nil, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var env EnvironmentInfo
	if err := json.NewDecoder(resp.Body).Decode(&env); err != nil {
		return nil, fmt.Errorf("decode environment: %w", err)
	}
	return &env, nil
}

func (s *EnvironmentsService) Create(ctx context.Context, reqBody CreateRequest) (*EnvironmentInfo, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	req, err := s.t.newRequest(ctx, http.MethodPost, "/v1/environments", nil, reqBody)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var env EnvironmentInfo
	if err := json.NewDecoder(resp.Body).Decode(&env); err != nil {
		return nil, fmt.Errorf("decode environment: %w", err)
	}
	return &env, nil
}

// Drain is phase-1 only: it quiesces the guest workload and moves the vm to
// "draining" so you can inspect its output, but it does not destroy the vm.
// Call Destroy afterward to remove it - a drained vm left undestroyed keeps
// running (and billing) forever.
func (s *EnvironmentsService) Drain(ctx context.Context, vmID string) (*EnvironmentInfo, error) {
	return s.action(ctx, vmID, "drain")
}

// Fork creates a new environment seeded from an existing one and returns the
// new environment. when opts.ReuseSnapshotID is empty the server snapshots the
// source first; otherwise it reuses the named snapshot. it does not use the
// shared action helper because fork sends a request body.
func (s *EnvironmentsService) Fork(ctx context.Context, vmID string, opts ForkOptions) (*EnvironmentInfo, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	if vmID == "" {
		return nil, errors.New("vm id is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID)
	values := url.Values{}
	values.Set("action", "fork")
	req, err := s.t.newRequest(ctx, http.MethodPost, path, values, opts)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var env EnvironmentInfo
	if err := json.NewDecoder(resp.Body).Decode(&env); err != nil {
		return nil, fmt.Errorf("decode environment: %w", err)
	}
	return &env, nil
}

func (s *EnvironmentsService) RotateToken(ctx context.Context, vmID string) error {
	if s == nil || s.t == nil {
		return errors.New("environments service is not configured")
	}
	if vmID == "" {
		return errors.New("vm id is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID)
	values := url.Values{}
	values.Set("action", "rotate-token")
	req, err := s.t.newRequest(ctx, http.MethodPost, path, values, nil)
	if err != nil {
		return err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return err
	}
	if err := CheckResponse(resp); err != nil {
		return err
	}
	return nil
}

func (s *EnvironmentsService) Destroy(ctx context.Context, vmID string) error {
	if s == nil || s.t == nil {
		return errors.New("environments service is not configured")
	}
	if vmID == "" {
		return errors.New("vm id is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID)
	req, err := s.t.newRequest(ctx, http.MethodDelete, path, nil, nil)
	if err != nil {
		return err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return err
	}
	if err := CheckResponse(resp); err != nil {
		return err
	}
	return nil
}

func (s *EnvironmentsService) action(ctx context.Context, vmID, action string) (*EnvironmentInfo, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	if vmID == "" {
		return nil, errors.New("vm id is required")
	}
	if action == "" {
		return nil, errors.New("action is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID)
	values := url.Values{}
	values.Set("action", action)
	req, err := s.t.newRequest(ctx, http.MethodPost, path, values, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var env EnvironmentInfo
	if err := json.NewDecoder(resp.Body).Decode(&env); err != nil {
		return nil, fmt.Errorf("decode environment: %w", err)
	}
	return &env, nil
}

// Computer relays one computer-use action to the environment's desktop and
// returns the result, usually carrying a screenshot. It requires an
// environment booted from a desktop image; on any other image the server
// answers 503 with a reason. Screenshots ride back base64-encoded, so the
// stream client with no overall timeout carries the call.
func (s *EnvironmentsService) Computer(ctx context.Context, vmID string, action ComputerActionRequest) (*ComputerActionResponse, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	if vmID == "" {
		return nil, errors.New("vm id is required")
	}
	if action.Action == "" {
		return nil, errors.New("action is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID) + "/computer"
	req, err := s.t.newRequest(ctx, http.MethodPost, path, nil, action)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.doStream(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var res ComputerActionResponse
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return nil, fmt.Errorf("decode computer response: %w", err)
	}
	return &res, nil
}

// ComputerDisplay reports whether the environment has a live display and at
// what geometry. Up is false with a reason on an image with no desktop.
func (s *EnvironmentsService) ComputerDisplay(ctx context.Context, vmID string) (*ComputerDisplay, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("environments service is not configured")
	}
	if vmID == "" {
		return nil, errors.New("vm id is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID) + "/computer"
	req, err := s.t.newRequest(ctx, http.MethodGet, path, nil, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var res ComputerDisplay
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return nil, fmt.Errorf("decode display report: %w", err)
	}
	return &res, nil
}

File: sdks/typescript/src/environments.ts
import type { CallOptions, Transport } from "./transport.js";
import type {
  ComputerAction,
  ComputerDisplay,
  ComputerResult,
  ComputerToolResult,
  CreateRequest,
  EnvironmentInfo,
  Event,
  ExecRequest,
  ExecResult,
  ForkOptions,
  ToolResultBlock,
} from "./types.js";
import { FuseError, isFuseApiError } from "./errors.js";
import { requireArg } from "./validate.js";
import { streamEvents } from "./events.js";

/** The server's max page size; list() requests this internally so it walks
 * every page in as few round trips as possible. */
const MAX_PAGE_LIMIT = 200;

/** Filters (and pagination) for environments.list / environments.listPage. */
export interface ListEnvironmentsOptions {
  taskId?: string;
  state?: string;
  hostId?: string;
  /** Page size (server default 50, max 200). Only consulted by listPage —
   * list() always walks every page, so it ignores this and requests the
   * server's max page size. */
  limit?: number;
  /** Opaque cursor from a previous listPage() call's nextCursor. */
  cursor?: string;
}

interface EnvironmentList {
  environments: EnvironmentInfo[];
  next_cursor?: string;
}

/** One page of environments.listPage, plus the cursor to fetch the next one. */
export interface EnvironmentPage {
  environments: EnvironmentInfo[];
  /** Empty/undefined once there are no more results. */
  nextCursor?: string;
}

/** EnvironmentsService manages microVM lifecycle. */
export class EnvironmentsService {
  constructor(private readonly t: Transport) {}

  /** List tracked environments, optionally filtered. Transparently walks
   * every result page. For explicit single-page control (e.g. a cursor from
   * a previous call), use listPage. */
  async list(
    options: ListEnvironmentsOptions = {},
    opts: CallOptions = {},
  ): Promise<EnvironmentInfo[]> {
    const out: EnvironmentInfo[] = [];
    let cursor = options.cursor;
    for (;;) {
      const page = await this.listPage(
        { ...options, limit: MAX_PAGE_LIMIT, cursor },
        opts,
      );
      out.push(...page.environments);
      if (!page.nextCursor) break;
      cursor = page.nextCursor;
    }
    return out;
  }

  /** List one page of tracked environments, optionally filtered. */
  async listPage(
    options: ListEnvironmentsOptions = {},
    opts: CallOptions = {},
  ): Promise<EnvironmentPage> {
    const out = await this.t.json<EnvironmentList>("GET", "/v1/environments", {
      query: {
        task_id: options.taskId,
        state: options.state,
        host_id: options.hostId,
        limit: options.limit ? String(options.limit) : undefined,
        cursor: options.cursor,
      },
      signal: opts.signal,
    });
    return { environments: out.environments ?? [], nextCursor: out.next_cursor };
  }

  /** Fetch a single environment by VM id. */
  async get(vmId: string, opts: CallOptions = {}): Promise<EnvironmentInfo> {
    requireArg(vmId, "vm id");
    return this.t.json<EnvironmentInfo>(
      "GET",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { signal: opts.signal },
    );
  }

  /** Provision a new environment. Blocks until running or creation fails. */
  async create(body: CreateRequest, opts: CallOptions = {}): Promise<EnvironmentInfo> {
    return this.t.json<EnvironmentInfo>("POST", "/v1/environments", {
      body,
      signal: opts.signal,
    });
  }

  /** Fork an environment into a new one; returns the new environment. */
  async fork(
    vmId: string,
    body: ForkOptions = {},
    opts: CallOptions = {},
  ): Promise<EnvironmentInfo> {
    requireArg(vmId, "vm id");
    return this.t.json<EnvironmentInfo>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { query: { action: "fork" }, body, signal: opts.signal },
    );
  }

  /**
   * Run a command inside a running environment's guest and return its exit code
   * with stdout and stderr kept separate.
   *
   * A non-zero exit_code is resolved, not thrown: the command ran and failed.
   * A thrown error means the command could not be run at all.
   *
   * Exec requires the master token.
   *
   * @example
   * const out = await client.environments.exec(id, { cmd: ["ls", "-l"] });
   * if (out.exit_code !== 0) console.error(out.stderr);
   */
  async exec(
    vmId: string,
    body: ExecRequest,
    opts: CallOptions = {},
  ): Promise<ExecResult> {
    requireArg(vmId, "vm id");
    const hasCmd = (body.cmd?.length ?? 0) > 0;
    const hasShell = (body.shell ?? "") !== "";
    if (!hasCmd && !hasShell) {
      throw new FuseError("one of cmd or shell is required");
    }
    if (hasCmd && hasShell) {
      throw new FuseError("cmd and shell are mutually exclusive");
    }
    return this.t.json<ExecResult>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { query: { action: "exec" }, body, signal: opts.signal },
    );
  }

  /**
   * Relay one computer-use action to the environment's desktop and return the
   * result, usually carrying a base64 PNG screenshot. Requires an environment
   * booted from a desktop image; on any other image the server answers 503
   * with a reason.
   *
   * @example
   * const res = await client.environments.computer(id, {
   *   action: "left_click",
   *   coordinate: [512, 384],
   * });
   */
  async computer(
    vmId: string,
    action: ComputerAction,
    opts: CallOptions = {},
  ): Promise<ComputerResult> {
    requireArg(vmId, "vm id");
    requireArg(action?.action, "action");
    return this.t.json<ComputerResult>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}/computer`,
      { body: action, signal: opts.signal },
    );
  }

  /**
   * Execute one computer tool_use against the environment and shape the
   * answer for the Messages API: the tool_use block's `input` goes in
   * verbatim (it is already the wire shape the computer endpoint takes), and
   * content blocks for the tool_result come out.
   *
   * An action the environment refuses (a malformed coordinate, a display
   * that is not up) resolves to `is_error` content rather than throwing: the
   * agent loop should feed it to the model, which is the party that can
   * correct it. A throw is reserved for failures the model cannot fix —
   * transport, auth, an unknown environment.
   *
   * @example
   * for (const block of msg.content) {
   *   if (block.type === "tool_use" && block.name === "computer") {
   *     const result = await client.environments.computerToolResult(id, block.input);
   *     toolResults.push({ type: "tool_result", tool_use_id: block.id, ...result });
   *   }
   * }
   */
  async computerToolResult(
    vmId: string,
    input: unknown,
    opts: CallOptions = {},
  ): Promise<ComputerToolResult> {
    requireArg(vmId, "vm id");
    const action = (input as { action?: string } | null)?.action ?? "";
    if (action === "") {
      return {
        is_error: true,
        content: [{ type: "text", text: "tool input names no action" }],
      };
    }
    let res: ComputerResult;
    try {
      res = await this.t.json<ComputerResult>(
        "POST",
        `/v1/environments/${encodeURIComponent(vmId)}/computer`,
        { body: input, signal: opts.signal },
      );
    } catch (err) {
      if (
        isFuseApiError(err) &&
        (err.status === 400 || err.status === 409 || err.status === 503)
      ) {
        return {
          is_error: true,
          content: [{ type: "text", text: err.message || "the computer action failed" }],
        };
      }
      throw err;
    }
    const content: ToolResultBlock[] = [];
    if (res.output) content.push({ type: "text", text: res.output });
    if (res.screenshot) {
      content.push({
        type: "image",
        source: {
          type: "base64",
          media_type: "image/png",
          data: res.screenshot,
        },
      });
    }
    if (content.length === 0) content.push({ type: "text", text: "done" });
    return { content };
  }

  /** Report whether the environment has a live display and at what geometry.
   * up is false with a reason on an image with no desktop. */
  async computerDisplay(vmId: string, opts: CallOptions = {}): Promise<ComputerDisplay> {
    requireArg(vmId, "vm id");
    return this.t.json<ComputerDisplay>(
      "GET",
      `/v1/environments/${encodeURIComponent(vmId)}/computer`,
      { signal: opts.signal },
    );
  }

  /**
   * Phase-1 only: quiesces the guest workload and moves the environment to
   * "draining" so you can inspect its output, but does not destroy it. Call
   * destroy() afterward to remove it - a drained environment left
   * undestroyed keeps running (and billing) forever.
   */
  async drain(vmId: string, opts: CallOptions = {}): Promise<EnvironmentInfo> {
    requireArg(vmId, "vm id");
    return this.t.json<EnvironmentInfo>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { query: { action: "drain" }, signal: opts.signal },
    );
  }

  /** Rotate the per-VM credentials and push them to the guest. */
  async rotateToken(vmId: string, opts: CallOptions = {}): Promise<void> {
    requireArg(vmId, "vm id");
    await this.t.noContent("POST", `/v1/environments/${encodeURIComponent(vmId)}`, {
      query: { action: "rotate-token" },
      signal: opts.signal,
    });
  }

  /** Forcefully destroy an environment. Idempotent. */
  async destroy(vmId: string, opts: CallOptions = {}): Promise<void> {
    requireArg(vmId, "vm id");
    await this.t.noContent("DELETE", `/v1/environments/${encodeURIComponent(vmId)}`, {
      signal: opts.signal,
    });
  }

  /**
   * Open the SSE event stream for an environment. The returned promise rejects
   * with a FuseApiError on a connect-time failure (e.g. not_found); on success
   * it resolves to an AsyncIterable that yields each Event and ends after a
   * terminal-state event. Pass `opts.signal` to cancel; there is no built-in
   * timeout on the stream.
   *
   * @example
   * for await (const ev of await client.environments.events(id)) {
   *   console.log(ev.state);
   * }
   */
  async events(vmId: string, opts: CallOptions = {}): Promise<AsyncIterable<Event>> {
    requireArg(vmId, "vm id");
    return streamEvents(this.t, vmId, opts.signal);
  }
}
